File: Stack/0255. Verify Preorder Sequence in Binary Search Tree/0255.py
class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        stack=[]
        currRoot=-1

        for node in preorder:
            while stack and node>stack[-1]:
                currRoot=stack.pop()

            if node<=currRoot:
                return False

            stack.append(node)
            
        return True

File: Stack/0255. Verify Preorder Sequence in Binary Search Tree/0255.cpp
#pragma GCC optimize('0fast', 'unroll-loops')

class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        stack<int>st;
        int currRoot=-1;

        for(int currNode : preorder){
            while(!st.empty() && currNode>st.top()){
                currRoot=st.top();
                st.pop();
            }
            if(currNode<=currRoot) return false;

            st.push(currNode);
        }
        return true;

    }
};